File: 남기석/redis/src/main/java/com/example/demo/infra/jwt/JwtTokenProvider.java
package com.example.demo.infra.jwt;



import com.example.demo.module.service.AccountService;
import io.jsonwebtoken.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;
import com.example.demo.module.dto.response.UserResponse;

import javax.annotation.PostConstruct;
import javax.servlet.http.HttpServletRequest;
import java.util.Base64;
import java.util.Date;


@Slf4j
@Component
@RequiredArgsConstructor
public class JwtTokenProvider {
    private static final long ACCESS_TOKEN_EXPIRE_TIME = 30 * 60 * 1000L;              // 30분
    private static final long REFRESH_TOKEN_EXPIRE_TIME = 7 * 24 * 60 * 60 * 1000L;    // 7일
    private final AccountService accountService;


    private String secretKey = "secretKey";
    @PostConstruct
    protected void init() {
        secretKey = Base64.getEncoder().encodeToString(secretKey.getBytes());
    }


    // 인증정보 기반으로 토큰 생성
    public UserResponse.TokenInfo generateToken(String username) {
        Claims claims = Jwts.claims().setSubject(username);
        Date date = new Date();
        // accessToken 생성
        String accessToken = Jwts.builder()
                .setClaims(claims)
                .setIssuedAt(date)
                .setExpiration(new Date(date.getTime() + ACCESS_TOKEN_EXPIRE_TIME))
                .signWith(SignatureAlgorithm.HS256,secretKey)
                .compact();

        //refresh Token생성
        String refreshToken = Jwts.builder()
                .setExpiration(new Date(date.getTime() + REFRESH_TOKEN_EXPIRE_TIME))
                .signWith(SignatureAlgorithm.HS256,secretKey)
                .compact();

        return UserResponse.TokenInfo.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .refreshTokenExpirationTime(REFRESH_TOKEN_EXPIRE_TIME)
                .build();
    }

    // 토큰 인증정보 조회
    public Authentication getAuth(String token) {
        //token에서 username정보 꺼낸 후
        UserDetails userDetails = accountService.loadUserByUsername(getUserPk(token));
        return new UsernamePasswordAuthenticationToken(userDetails, "", userDetails.getAuthorities());
    }

    // 토큰에서 회원 정보 추출
    public String getUserPk(String token) {
        return Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token).getBody().getSubject();
    }

    // 요청 헤더 중 X-AUTH-TOKEN헤더에서 토큰 가져온다
    public String getToken(HttpServletRequest req) {
        return req.getHeader("X-AUTH-TOKEN");
    }


    //토큰 유효성검사
    public boolean validateToken(String token) {
        try {
            Jws<Claims> claimsJws = Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token);
//            if (logoutList.contains(token)) {
//                return false;
//            }
            return !claimsJws.getBody().getExpiration().before(new Date());
        } catch (Exception e) {
            return false;
        }
    }
}

File: Jose/redis/src/main/kotlin/com/jose/redis/jwt/JwtTokenProvider.kt
package com.jose.redis.jwt

import com.jose.redis.enum.RoleType
import com.jose.redis.exception.jwtException.InvalidTokenException
import com.jose.redis.security.TokenInfo
import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.MalformedJwtException
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.SignatureException
import io.jsonwebtoken.UnsupportedJwtException
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.User
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.stereotype.Component
import java.util.Arrays
import java.util.Date
import java.util.stream.Collectors
import javax.naming.AuthenticationException

@Component
class JwtTokenProvider {
    private val ACCESS_TOKEN_VALID_TIME: Long = (1000 * 60 * 60 * 24) // 하루
    private val REFRESH_TOKEN_VALID_TIME: Long = (1000 * 60 * 60 * 24 * 7) // 일주일
    private val BEARER_TYPE = "Bearer "
    @Value("\${jwt.secret}")
    lateinit var secretKey: String

    fun createJwtAccessToken(email: String): TokenInfo {
        val now = Date()
        val expiration = Date(now.time + ACCESS_TOKEN_VALID_TIME)
        var accessToken: String = Jwts.builder()
            .claim("email", email)
            .claim("ROLE_", RoleType.USER)
            .setIssuedAt(now)
            .setExpiration(expiration)
            .signWith(SignatureAlgorithm.HS256, secretKey)
            .compact()
        accessToken = BEARER_TYPE + accessToken
        return TokenInfo(accessToken, expiration.time - now.time)
    }

    fun createJwtRefreshToken(email: String): TokenInfo {
        val now = Date()
        val expiration = Date(now.time + REFRESH_TOKEN_VALID_TIME)
        var refreshToken: String = Jwts.builder()
            .claim("email", email)
            .claim("ROLE_", RoleType.USER)
            .setIssuedAt(now)
            .setExpiration(expiration)
            .signWith(SignatureAlgorithm.HS256, secretKey)
            .compact()
        refreshToken = BEARER_TYPE + refreshToken
        return TokenInfo(refreshToken, expiration.time - now.time)
    }

    fun validateToken(token: String): Boolean {
        try {
            Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token)
            return true
        } catch (e: SecurityException) {
            throw InvalidTokenException("JWT에러 Security")
        } catch (e: MalformedJwtException) {
            throw InvalidTokenException("JWT에러 Malformed")
        } catch (e: ExpiredJwtException) {
            throw InvalidTokenException("만료된 JWT 토큰입니다.")
        } catch (e: UnsupportedJwtException) {
            throw InvalidTokenException("지원되지 않는 JWT 토큰입니다.")
        } catch (e: IllegalArgumentException) {
            throw InvalidTokenException("잘못된 JWT 헤더 입니다.")
        } catch (e: SignatureException) {
            throw InvalidTokenException("잘못된 JWT 서명 입니다.")
        }
    }

    fun getAuthentication(token: String): Authentication {
        val claims = parseClaims(token)
        try {
            claims.get("email")
        } catch (e: Exception) {
            throw AuthenticationException("JWT 토큰에 email이 없습니다.")
        }
        val authorities: Collection<GrantedAuthority> =
            Arrays.stream(claims.get("ROLE_").toString().split(",").toTypedArray())
                .map { role: String -> SimpleGrantedAuthority(role) }
                .collect(Collectors.toList())
        val userDetails: UserDetails = User(claims.get("email").toString(), "", authorities)
        return UsernamePasswordAuthenticationToken(userDetails, "", userDetails.authorities)
    }

    fun parseClaims(token: String?): Claims {
        return try {
            Jwts.parser().setSigningKey(secretKey).parseClaimsJws(token).body
        } catch (e: ExpiredJwtException) {
            e.claims
        }
    }

    fun getExpiration(token: String): Long {
        val expiration: Date =
            Jwts.parser().setSigningKey(secretKey).parseClaimsJws(resolveToken(token)).body.expiration
        val now = Date()
        return expiration.time - now.time
    }

    fun resolveToken(token: String): String? {
        if (token.startsWith(BEARER_TYPE)) {
            return token.replace("Bearer ", "")
        }
        return null
    }
}
